Capture fx-graph-cache-key in PyTorch trace

shengfukevin opened the pull request:
Summary:
Since PT2 compiler generated Triton kernels are generated on-fly, it is a black box for workload performance simulator from AI system Co-Design team. It is impossible to get the performance projection for these kernels.

The first step to get inside into Triton kernels is to get the source code. FX graph remote cache contains the source code of triton kernel. In PyTorch trace, we want to collect the cache key. These traces got processed in Durin, user can then retrieve triton source code from gpu_kernel_stats in Durin based on the cache key.

This DIFF is to pass fx graph cache key in Inductor meta data, and then it will be captured in PyTorch trace.

Test Plan: buck2 run mode/opt caffe2/test/inductor:profiler -- caffe2.test.inductor.test_profiler.DynamoProfilerTests.test_pt2_triton_fx_graph_cache_key

Differential Revision: D71452637


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

shengfukevin commented:
@pytorchbot label "topic: not user facing"

jamesjwu reviewed (changes requested):
Looks like theres a bunch of failures and stuff you'd need to thread through to have this work. 

Some questions: 
- How exactly does knowing the fx graph cache key help find the source code of the triton? Why not use, say the triton cache key? Or figure out a way to log the triton source code directly? 
- Changing the `inductor_meta` affects the autotune cache key (I remember @masnesral running into this when adding compile ID). Have you tested or reasoned about that in this case, autotuning won't cache miss more from your change?

eellison reviewed (commented):
please revert all the formatting changes or submit as different pr.

oulgen commented:
Cache key is already captured on the tlparse and scuba logging? Why is that not enough?

jansel reviewed (changes requested):
Failing tests?

shengfukevin committed:
Capture fx-graph-cache-key in PyTorch trace (#149580)

Summary:

Since PT2 compiler generated Triton kernels are generated on-fly, it is a black box for workload performance simulator from AI system Co-Design team. It is impossible to get the performance projection for these kernels.

The first step to get inside into Triton kernels is to get the source code. FX graph remote cache contains the source code of triton kernel. In PyTorch trace, we want to collect the cache key. These traces got processed in Durin, user can then retrieve triton source code from gpu_kernel_stats in Durin based on the cache key.

This DIFF is to pass fx graph cache key in Inductor meta data, and then it will be captured in PyTorch trace.

Test Plan: buck2 run mode/opt caffe2/test/inductor:profiler -- caffe2.test.inductor.test_profiler.DynamoProfilerTests.test_pt2_triton_fx_graph_cache_key

Differential Revision: D71452637

briancoutinho commented:
> Cache key is already captured on the tlparse and scuba logging? Why is that not enough?

@oulgen that's adding multiple hops for identifying the FX graph
1) Internally we process the GPU traces to obtain statistics per GPU kernel (https://www.internalfb.com/wiki/Durin/)
If we have the key here we can add this to our kernel leveldataset.
2) Then add a utility that downloads the cached metadata by the fx graph key.

The alternative is to associate two different tables linked via the training job. Our kernel level dataset only has the hash of the triton kernel today . We would have to download all the cached FX objects and then match the triton kernels from there to the kernel.

We discussed this last year and the format of the cache and table are not easy to interpret.

briancoutinho commented:
> Changing the inductor_meta affects the autotune cache key (I remember @masnesral running into this when adding compile ID). Have you tested or reasoned about that in this case, autotuning won't cache miss more from your change?

Yes we need to ensure we are not changing the autotune cache or having sideeffects.
Is there some way to pass down metadata to the runtime component without it being a factor in the caching?

shengfukevin commented:
> Cache key is already captured on the tlparse and scuba logging? Why is that not enough?

Is fx_graph_cache_key logged to tlparse through CompileEventLogger? I searched the Inductor code base, do not see it.

Which scuba table has fx_graph_cache_key? 

Let's say we have the fx_graph_cache_key somewhere, how do I know a triton kernel is included in that fx graph cache (assume a model may have a few fx graphs) ? Do I need to download all fx graph caches and search through it?

shengfukevin commented:
> Looks like theres a bunch of failures and stuff you'd need to thread through to have this work.
> 
> Some questions:
> 
> * How exactly does knowing the fx graph cache key help find the source code of the triton? Why not use, say the triton cache key? Or figure out a way to log the triton source code directly?

I developed a tool query_fx_graph_remtoe_cache in this DIFF https://www.internalfb.com/diff/D69341750. It uses fx graph cache key to download the cache, and get the source code (one example) https://fburl.com/manifold/vv5vk6mr 
There are probably other ways to get that, but this solution seems very clean to me. 

Also I think the availability of fx graph remote cache should be high since user depends on it to speed up the compilation. Other logging mechanism may be only enabled for debugging. We want to have a good coverage for fleet wide models.

> * Changing the `inductor_meta` affects the autotune cache key (I remember @masnesral running into this when adding compile ID). Have you tested or reasoned about that in this case, autotuning won't cache miss more from your change?

This is a good question, what is your suggestion to pass fx_graph_cache_key to class CachingAutotuner?

shengfukevin commented:
This is a note for downloading fx graph remote cache and deserialized it: 

https://docs.google.com/document/d/1irhgGbTUiRW9xiitepwVDuVXciHIPT0OD0qcxOVND0Y/edit?usp=sharing

jamesjwu commented:
> This is a good question, what is your suggestion to pass fx_graph_cache_key to class CachingAutotuner?

@masnesral might have a good suggestion on how he's doing it for compile ID

masnesral commented:
> > This is a good question, what is your suggestion to pass fx_graph_cache_key to class CachingAutotuner?
> 
> @masnesral might have a good suggestion on how he's doing it for compile ID

Yeah, I found out the hard way that you don't want to stick these kind of identifiers in the inductor_meta because it breaks the Triton caching. For the compileId, it's not very tricky. Since we have the compileId available at the point we get the CachingAutotuner back from the triton-compile-subprocess, we opted to set it at that point:

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/async_compile.py#L330
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/async_compile.py#L352

I'm not sure that something similar works for trying to store the fx graph cache key though.
